crates/sui-source-validation/src/lib.rs

```diff
@@ -396,3 +467,306 @@ fn local_modules(
+                );
+                package_version_map.insert(*package, (toolchain_version, vec![local_unit.clone()]));
+            }
+        }
+    }
+
+    let mut units = vec![];
+    // Iterate over compiled units, and check if they need to be recompiled and replaced by a prior compiler's output.
+    for (package, (toolchain_version, local_units)) in package_version_map {
+        if toolchain_version.compiler_version == CURRENT_COMPILER_VERSION {
+            let local_units: Vec<_> = local_units.iter().map(|u| (package, u.clone())).collect();
+            units.extend(local_units);
+            continue;
+        }
+
+        if local_units.is_empty() {
+            bail!("Expected one or more modules, but none found");
+        }
+        let package_root = SourcePackageLayout::try_find_root(&local_units[0].source_path)?;
+        let install_dir = tempfile::tempdir()?; // place compiled packages in this temp dir, don't pollute this packages build dir
+        download_and_compile(
+            package_root.clone(),
+            &install_dir,
+            &toolchain_version,
+            &package,
+        )?;
+
+        let compiled_unit_paths = vec![package_root.clone()];
+        let compiled_units = find_filenames(&compiled_unit_paths, |path| {
+            extension_equals(path, MOVE_COMPILED_EXTENSION)
+        })?;
+        let build_path = install_dir
+            .path()
+            .join(CompiledPackageLayout::path(&CompiledPackageLayout::Root))
+            .join(package.as_str());
+        debug!("build path is {}", build_path.display());
+
+        // Add all units compiled with the previous compiler.
+        for bytecode_path in compiled_units {
+            info!("bytecode path {bytecode_path}, {package}");
+            let local_unit = decode_bytecode_file(build_path.clone(), &package, &bytecode_path)?;
+            units.push((package, local_unit))
+        }
+    }
+    Ok(units)
+}
+
+fn download_and_compile(
+    root: PathBuf,
+    install_dir: &TempDir,
+    ToolchainVersion {
+        compiler_version,
+        edition,
+        flavor,
+    }: &ToolchainVersion,
+    dep_name: &Symbol,
+) -> anyhow::Result<()> {
+    let dest_dir = PathBuf::from_iter([&*MOVE_HOME, "binaries"]); // E.g., ~/.move/binaries
+    let dest_version = dest_dir.join(compiler_version);
+    let mut dest_canonical_binary = dest_version.clone();
+    dest_canonical_binary.extend(["target", "release", CANONICAL_BINARY_NAME]);
+
+    if !dest_canonical_binary.exists() {
+        // Check the platform and proceed if we can download a binary. If not, the user should follow error instructions to sideload the binary.
+        let platform = detect_platform(&root, compiler_version, &dest_canonical_binary)?;
+        // Download if binary does not exist.
+        let url = format!("https://github.com/MystenLabs/sui/releases/download/mainnet-v{}/sui-mainnet-v{}-{}.tgz", compiler_version, compiler_version, platform);
```

Comment:
Is it going to be a problem that we use mainnet releases always? (what if someone is building a package for testnet and their dependency was built with an older version of the testnet toolchain but that still haven't reached mainnet yet?)

Reply:
Hmm, does it happen that we release a testnet version without a corresponding mainnet version and cut testnet for the next version?

I'll address this in follow up--I think the best past forward is to give an error that the mainnet version doesn't exist, and a user can download and side load the binary.

We could pin to the testnet binary, but that feels somehow strange, and I'm not sure if there's a chance of testnet vs mainnet binaries diverging either, so my gut says we should stick to mainnet releases or bust?

Reply:
> Hmm, does it happen that we release a testnet version without a corresponding mainnet version and cut testnet for the next version?

Yes this can happen unfortunately -- generally `mainnet` leads `testnet` by one version, so if you are building on testnet, you may be one version ahead.

I think it's reasonable to treat `mainnet` as the "stable" revision -- perhaps if you don't find a mainnet version, you could try to look for the testnet version as a backup under the assumption that you couldn't find the version because it's "too new"?